### PR TITLE
Hw 3

### PR DESCRIPTION
**Pull request by MaliukDaria**

### CODE REVIEW

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> findByEmailQuery = session.createQuery(
+                    "FROM User WHERE email = :userEmail", User.class)
+                    .setParameter("userEmail", email);
+            return Optional.ofNullable(findByEmailQuery.uniqueResult());
```
Comment (andriylagun) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: we can use findByEmailQuery.uniqueResultOptional();

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+public class UserServiceImpl implements UserService {
+    @Inject
+    private UserDao userDao;
+
+    @Override
+    public User add(User user) {
+        user.setSalt(HashUtil.getSalt());
+        user.setPassword(HashUtil.hashPassword(user.getPassword(), user.getSalt()));
+        return userDao.add(user);
+    }
```
Comment (andriylagun) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java:22]: I think maybe better will be setSalt and encrypt password in register method, but not sure

File: src/main/java/com/dev/cinema/MainApp.java [hunk 3]
```diff
@@ -1,20 +1,24 @@
...
+import com.dev.cinema.service.UserService;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.util.List;
 
 public class MainApp {
     private static Injector injector = Injector.getInstance("com.dev.cinema");
 
-    public static void main(String[] args) {
+    public static void main(String[] args) throws AuthenticationException {
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/MainApp.java]: давай обгорнемо помилку в try-catch

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> user = userService.findByEmail(email);
+        if (user.isPresent()
+                && user.get().getPassword()
+                .equals(HashUtil.hashPassword(password, user.get().getSalt()))) {
+            return user.get();
+        } else {
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: do we need else block here?

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java (hunk 1 above)
Comment (MaliukDaria) [reply src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: done

File: src/main/java/com/dev/cinema/MainApp.java (hunk 3 above)
Comment (MaliukDaria) [reply src/main/java/com/dev/cinema/MainApp.java]: done

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java (hunk 4 above)
Comment (MaliukDaria) [reply src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: done

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**